pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -352,3 +378,62 @@ func isMaster(machine mapiv1.Machine, client client.Client) bool {
 	}
 	return false
 }
+
+// checkAndDecrement checks if the provided MachineDisruptionBudget allows any disruption
+func (r *ReconcileMachineHealthCheck) checkAndDecrement(machineName string, mdb *healthcheckingv1alpha1.MachineDisruptionBudget, maxDisruptedMachinSize int) error {
+	if mdb.Status.ObservedGeneration < mdb.Generation {
+		return fmt.Errorf("the machine disruption budget %s is still being processed by the server", mdb.Name)
+	}
+	if mdb.Status.MachineDisruptionsAllowed < 0 {
+		return fmt.Errorf("the machine disruption budget %s MachineDisruptionsAllowed is negative", mdb.Name)
+	}
+	if len(mdb.Status.DisruptedMachines) > maxDisruptedMachinSize {
+		return fmt.Errorf("the machine disruption budget %s DisruptedMachines map too big - too many deletions not confirmed by MDB controller", mdb.Name)
+	}
+	if mdb.Status.MachineDisruptionsAllowed == 0 {
+		return fmt.Errorf("cannot delete machine as it would violate the machine's disruption budget %s", mdb.Name)
+	}
+
+	mdb.Status.MachineDisruptionsAllowed--
+	if mdb.Status.DisruptedMachines == nil {
+		mdb.Status.DisruptedMachines = make(map[string]metav1.Time)
+	}
+	// MachineHealthCheck controller needs to inform the MDB controller that it is about to delete a machine
+	// so it should not consider it as available in calculations when updating MachineDisruptions allowed.
+	// If the machine is not deleted within a reasonable time limit MDB controller will assume that it won't
+	// be deleted at all and remove it from DisruptedMachines map.
+	mdb.Status.DisruptedMachines[machineName] = metav1.Time{Time: time.Now()}
```

Comment:
I'm less keen about this part but perhaps it unavoidable when not using the Ready condition.
It would have been nice if MDB could automagically exclude "unhealthy" nodes from its calculations.

Could/should we have MDB consume the fencing config map?

Reply:
I think we have here two questions:

- all this mdb.Status.DisruptedMachines does not regarding health and unhealthy machines, but regarding the delay in the deletion of the machine so we do want to take into account machines that will soon be deleted, and if it does not be deleted after the specified timeout we start to take them in the account again
- I will use the node conditions config map under the MDB in the next iteration